packages/i18n/src/index.js

```diff
@@ -34,7 +34,9 @@ export function setLocaleData( localeData = { '': {} }, domain = 'default' ) {
 		} );
 	}
 
-	i18n.options.locale_data[ domain ] = localeData;
+	i18n.options.locale_data[ domain ] = i18n.options.locale_data.hasOwnProperty( domain )
```

Comment:
Okay, one more question 😄 

Do we care about mutating `i18n.options.locale_data` ?

Could skip an object clone to drop the first `{}` argument of `Object.assign` and assign directly.

Maybe as:

```js
if ( i18n.options.locale_data.hasOwnProperty( domain ) ) {
	Object.assign( i18n.options.locale_data[ domain ], localeData )
} else {
	i18n.options.locale_data[ domain ] = localeData;
}
```

Direct assignment with `=` is fine too, since `Object.assign` returns the resulting target object.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign

Or: We could simplify this to always assign into the `{}`, even if `! i18n.options.locale_data.hasOwnProperty( domain )`, since it's fine if anything but the first argument of `Object.assign` is `undefined`:

```js
i18n.options.locale_data[ domain ] = Object.assign( {}, i18n.options.locale_data[ domain ], localeData );
```

Reply:
Hmm I'm fine either way.  I wasn't sure if we should mutate the object directly or not but in my case its likely a force of habit working with react/redux that I just merge into a new object.  So mutation should be okay here.
 
I _like_ your last example so I'll roll with it.